components/tidb_query_normal_expr/src/builtin_time.rs

```diff
@@ -557,6 +557,39 @@ impl ScalarFunc {
         let time = Time::from_days(ctx, days)?;
         Ok(Some(Cow::Owned(time)))
     }
+
+    #[inline]
+    pub fn make_date<'a>(
+        &self,
+        ctx: &mut EvalContext,
+        row: &[Datum],
+    ) -> Result<Option<Cow<'a, Time>>> {
+        let mut year = try_opt!(self.children[0].eval_int(ctx, row)) as i64;
+        let mut day = try_opt!(self.children[1].eval_int(ctx, row)) as i64;
+        if day <= 0 || year < 0 || year > 9999 {
+            return Ok(None);
+        }
+        if year < 70 {
+            year += 2000;
+        } else if year < 100 {
+            year += 1900;
+        }
+        year -= 1;
+        let d4 = year / 4;
+        let d100 = year / 100;
+        let d400 = year / 400;
+        let leap = d4 - d100 + d400;
+        day = day + leap + year * 365 + 365;
+        if day > 366 * 9999 || day <= 0 {
```

Comment:
It seems `day <= 0` looks unnecessary?

Reply:
I think that variable `day` could overflow in line 582

Reply:
Emm... You're right. It can overflow. It's better we handle it.
But I recommend using `checked_add`.

Reply:
I've moved `day > 366 * 9999` to the first `if`. Now all numbers should be low during computation. I've also added additional test cases.